test/e2e/framework/job/wait.go

```diff
@@ -82,14 +83,33 @@ func waitForJobPodsInPhase(ctx context.Context, c clientset.Interface, ns, jobNa
 // both conformance CI jobs with GA-only features and e2e CI jobs with all default-enabled features.
 // So, we need to skip "Complete" condition reason verifications in the e2e conformance test cases.
 func WaitForJobComplete(ctx context.Context, c clientset.Interface, ns, jobName string, reason *string, completions int32) error {
-	if err := wait.PollUntilContextTimeout(ctx, framework.Poll, JobTimeout, false, func(ctx context.Context) (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
+	// This function is called by HandleRetry, which will retry
+	// on transient API errors or stop polling in the case of other errors.
+	get := func(ctx context.Context) (*batchv1.Job, error) {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return nil, err
+		}
+		if isJobFailed(job) {
+			return nil, fmt.Errorf("job %q failed while waiting for its completion", klog.KObj(job))
```

Comment:
Can you use gomega.StopTrying here? That makes the intent clearer. It's also possible to attach the job, which will cause it to be dumped as YAML when rendering the failure.

```suggestion
			return nil, gomega.StopTrying("job failed while waiting for its completion").Attach("job", job))
```

Returning a custom error works because https://github.com/kubernetes/kubernetes/blob/1f9d2577da68290c0466c1404bbff6100e1e303a/test/e2e/framework/get.go#L94-L111 treats unknown errors as final, but IMHO `StopTrying` is better.

Reply:
Nice! I like it. Done!